src/java.base/share/classes/java/lang/String.java

```diff
@@ -737,138 +737,128 @@ public String(byte[] bytes, int offset, int length, Charset charset) {
      * disambiguate it against other similar methods of this class.
      */
     private String(Charset charset, byte[] bytes, int offset, int length) {
+        String str;
         if (length == 0) {
-            this.value = "".value;
-            this.coder = "".coder;
+            str = "";
         } else if (charset == UTF_8.INSTANCE) {
-            if (COMPACT_STRINGS) {
-                int dp = StringCoding.countPositives(bytes, offset, length);
-                if (dp == length) {
-                    this.value = Arrays.copyOfRange(bytes, offset, offset + length);
-                    this.coder = LATIN1;
-                    return;
+            str = utf8(bytes, offset, length);
+        } else if (charset == ISO_8859_1.INSTANCE) {
+            str = iso88591(bytes, offset, length);
+        } else if (charset == US_ASCII.INSTANCE) {
+            str = ascii(bytes, offset, length);
+        } else {
+            str = decode(charset, bytes, offset, length);
+        }
+        this(str);
+        if (COMPACT_STRINGS && coder == UTF16) {
+        	initCompressionFlag();
+        }
```

Comment:
This is redundant; the construction of `str` will have already initialized the compression flag.
Likewise, lines 1006-1008 are redundant.